### PR TITLE
proc: support position independent executables (PIE)

### PR DESCRIPTION
**Pull request by aarzilli**
```
proc: support position independent executables (PIE)

Support for position independent executables (PIE) on the native linux
backend, the gdbserver backend on linux and the core backend.
Also implemented in the windows native backend, but it can't be tested
because go doesn't support PIE on windows yet.

```

### CODE REVIEW

**Comment by aarzilli** [comment]
cc @hyangah @heschik I'm curious if this works on google pies.

**Comment by hyangah** [comment]
Thanks! It works for the google pies.

File: pkg/dwarf/frame/entries.go [hunk 1]
```diff
@@ -75,7 +76,7 @@ func (fdes FrameDescriptionEntries) FDEForPC(pc uint64) (*FrameDescriptionEntry,
 		}
 		return true
 	})
-	if idx == len(fdes) {
+	if idx == len(fdes) || !fdes[idx].Cover(pc) {
```
[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/dwarf/frame/entries.go]: It's not: https://play.golang.org/p/srQtNW_TKNJ

[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/dwarf/frame/entries.go]: Ah, well then this is due to a bug in the search function that's written. See this fixed version: https://play.golang.org/p/qg6MjICOEyf

[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/dwarf/frame/entries.go]: Alright, this change seemed wrong and redundant to me, and upon further investigation I found some odd things with the search code. The FDE search test is incorrect, the test data that is used assumed `entry.end` is an absolute address, which it is not... it's an offset from the begin address. Secondly, the search function itself is way more complicated than it needs to be.

This diff passes all tests, fixes the test, and is a lot cleaner:

```diff
diff --git a/pkg/dwarf/frame/entries.go b/pkg/dwarf/frame/entries.go
index 1c26891..73d4bbb 100644
--- a/pkg/dwarf/frame/entries.go
+++ b/pkg/dwarf/frame/entries.go
@@ -67,13 +67,7 @@ func (err *ErrNoFDEForPC) Error() string {
 // Returns the Frame Description Entry for the given PC.
 func (fdes FrameDescriptionEntries) FDEForPC(pc uint64) (*FrameDescriptionEntry, error) {
        idx := sort.Search(len(fdes), func(i int) bool {
-               if fdes[i].Cover(pc) {
-                       return true
-               }
-               if fdes[i].LessThan(pc) {
-                       return false
-               }
-               return true
+               return !fdes[i].LessThan(pc)
        })
        if idx == len(fdes) {
                return nil, &ErrNoFDEForPC{pc}
diff --git a/pkg/dwarf/frame/entries_test.go b/pkg/dwarf/frame/entries_test.go
index 577ad16..3eadd60 100644
--- a/pkg/dwarf/frame/entries_test.go
+++ b/pkg/dwarf/frame/entries_test.go
@@ -9,9 +9,9 @@ import (
 
 func TestFDEForPC(t *testing.T) {
        fde1 := &FrameDescriptionEntry{begin: 0, end: 49}
-       fde2 := &FrameDescriptionEntry{begin: 50, end: 99}
-       fde3 := &FrameDescriptionEntry{begin: 100, end: 200}
-       fde4 := &FrameDescriptionEntry{begin: 201, end: 245}
+       fde2 := &FrameDescriptionEntry{begin: 50, end: 49}
+       fde3 := &FrameDescriptionEntry{begin: 100, end: 99}
+       fde4 := &FrameDescriptionEntry{begin: 200, end: 45}
 
        frames := NewFrameIndex()
        frames = append(frames, fde1)
@@ -27,6 +27,15 @@ func TestFDEForPC(t *testing.T) {
        if node != fde1 {
                t.Fatal("Got incorrect fde")
        }
+
+       node, err = frames.FDEForPC(105)
+       if err != nil {
+               t.Fatal(err)
+       }
+
+       if node != fde3 {
+               t.Fatal("Got incorrect fde")
+       }
 }
 
 func BenchmarkFDEForPC(b *testing.B) {
```

[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/dwarf/frame/entries.go]: Yes, end is misnamed but this is still wrong: https://play.golang.org/p/0HD-JNm2240 (also consider that there could be a gap between two FDEs)

[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/dwarf/frame/entries.go]: You're right, after reading through the search algorithm docs again we must perform this check again.

Regardless, we should simplify the search predicate by returning the result of `fdes[i].LessThan` as opposed to explicitly returning a boolean around a conditional and fix the unit test to properly catch this error.

[the remark this answers is not in the document]
Comment (aarzilli) [reply pkg/dwarf/frame/entries.go]: I don't think LessThan is the right thing there, I think the right predicate is `fdes[i].Cover(pc) || fdes[i].Begin() >= pc`.

[the remark this answers is not in the document]
Comment (derekparker) [reply pkg/dwarf/frame/entries.go]: Yeah, I think you're right there, that satisfies the search invariant.

**Pull request merged**